File: 6.0/php/src/IpDbSearch.php
<?php

namespace Qqzeng\Ip;

use Exception;

class IpDbSearch
{
    private static $instance = null;
    
    private $data;
    private $geoispArr;
    
    const INDEX_START_INDEX = 0x30004;
    const END_MASK = 0x800000;
    const COMPL_MASK = ~self::END_MASK;
    const DB_FILE_NAME = 'qqzeng-ip-6.0-global.db';

    private function __construct()
    {
        $this->loadDb();
    }

    public static function getInstance()
    {
        if (self::$instance === null) {
            self::$instance = new self();
        }
        return self::$instance;
    }

    private function loadDb()
    {
        $dbPath = $this->findDbPath();
        if (!$dbPath) {
            throw new Exception("Fatal: Cannot find " . self::DB_FILE_NAME);
        }

        $this->data = file_get_contents($dbPath);
        if ($this->data === false) {
            throw new Exception("Failed to read database file");
        }

        if (strlen($this->data) < self::INDEX_START_INDEX) {
            throw new Exception("Invalid database file size");
        }

        // 节点数量 (小端序)
        // unpack 'V' 是 unsigned long (32bit, little endian)
        $nodeCount = unpack('V', substr($this->data, 0, 4))[1];
        
        $stringAreaOffset = self::INDEX_START_INDEX + $nodeCount * 6;
        
        if ($stringAreaOffset > strlen($this->data)) {
            throw new Exception("Invalid metadata");
        }

        // 解析字符串区
        $content = substr($this->data, $stringAreaOffset);
        $this->geoispArr = explode("\t", $content);
    }

    public function find($ip)
    {
        if (empty($ip)) {
            return "";
        }

        $ipLong = $this->fastParseIp($ip);
        if ($ipLong === false) {
            return "";
        }

        $prefix = ($ipLong >> 16) & 0xFFFF;
        $suffix = $ipLong & 0xFFFF;

        // 一级索引
        $record = $this->readInt24(4 + $prefix * 3);

        while (($record & self::END_MASK) !== self::END_MASK) {
            $bit = ($suffix >> 15) & 1;
            $offset = self::INDEX_START_INDEX + $record * 6 + $bit * 3;
            $record = $this->readInt24($offset);
            $suffix = ($suffix << 1) & 0xFFFF;
        }

        $index = $record & self::COMPL_MASK;
        if (isset($this->geoispArr[$index])) {
            return $this->geoispArr[$index];
        }
        return "";
    }

    private function readInt24($offset)
    {
        // PHP 字符串可以直接当数组访问字节
        // ord() 获取字节值
        // 大端序逻辑
        return (ord($this->data[$offset]) << 16) | 
               (ord($this->data[$offset + 1]) << 8) | 
               ord($this->data[$offset + 2]);
    }

    private function fastParseIp($ip)
    {
        // 极致性能使用 ip2long (C函数)
        // ip2long 返回的是有符号整型 (在32位系统上)，在64位系统上是int64
        // 我们需要由符号转无符号处理逻辑
        $long = ip2long($ip);
        if ($long === false) {
            return false;
        }
        // 确保是32位无符号
        return ($long < 0) ? ($long + 4294967296) : $long;
    }

    private function findDbPath()
    {
        $attempts = [
            __DIR__ . '/' . self::DB_FILE_NAME,
            getcwd() . '/' . self::DB_FILE_NAME,
            dirname(__DIR__, 2) . '/data/' . self::DB_FILE_NAME, // 假设在 src/Qqzeng/Ip
            dirname(__DIR__, 3) . '/data/' . self::DB_FILE_NAME,
            '../data/' . self::DB_FILE_NAME,
        ];

        foreach ($attempts as $path) {
            if (file_exists($path)) {
                return $path;
            }
        }
        return null;
    }
}

File: ip-classic-sdk/v6.0-db/netcore/IpDbSearch.cs
using System;
using System.IO;
using System.Text;
using System.Runtime.CompilerServices;

namespace qqzengIp
{
    /// <summary>
    /// qqzeng-ip 6.0 IP数据库解析类 (Safe/High-Performance)
    /// </summary>
    public sealed class IpDbSearch
    {
        private static readonly Lazy<IpDbSearch> _lazy = new Lazy<IpDbSearch>(() => new IpDbSearch());
        public static IpDbSearch Instance => _lazy.Value;

        private static byte[] _data = null!; 
        private static string[] _geoispArr = null!; 
        
        private const int IndexStartIndex = 0x30004; 
        private const int EndMask = 0x800000;
        private const int ComplMask = ~EndMask;
        private const string DbFileName = "qqzeng-ip-6.0-global.db";

        private IpDbSearch()
        {
            LoadDb();
        }

        private static void LoadDb()
        {
            try 
            {
                string dbPath = FindDbPath();
                if (string.IsNullOrEmpty(dbPath))
                {
                    dbPath = Path.Combine("../data", DbFileName);
                }

                if (!File.Exists(dbPath))
                {
                    throw new FileNotFoundException($"Fatal: Cannot find '{DbFileName}'");
                }

                _data = File.ReadAllBytes(dbPath);

                if (_data.Length < IndexStartIndex)
                {
                    throw new InvalidDataException("Invalid database file");
                }

                // NodeCount (小端序)
                // 仅初始化时运行一次，安全最重要
                int nodeCount = _data[0] | (_data[1] << 8) | (_data[2] << 16) | (_data[3] << 24);
                
                int stringAreaOffset = IndexStartIndex + nodeCount * 6;

                if (stringAreaOffset > _data.Length)
                {
                    throw new InvalidDataException("Invalid metadata");
                }

                _geoispArr = Encoding.UTF8.GetString(_data, stringAreaOffset, _data.Length - stringAreaOffset).Split('\t');
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Failed to initialize IpDbSearch.", ex);
            }
        }

        public string Find(string ip)
        {
            if (string.IsNullOrEmpty(ip)) return "";

            if (!FastParseIp(ip, out uint ipInt))
            {
                 return "";
            }

            return Find(ipInt);
        }

        /// <summary>
        /// 查询IP (Safe Optimized)
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public string Find(uint ipInt)
        {
            // 缓存到局部变量，避免由于静态字段跨方法访问造成的边界检查无法消除
            var data = _data; 
            
            int prefix = (int)(ipInt >> 16);
            int suffix = (int)(ipInt & 0xFFFF);

            // 一级索引
            // 4 + prefix * 3
            int idx = 4 + prefix * 3;
            // 手动 Check 边界让 JIT 知道后续不可能越界? 
            // 在热点代码中，我们直接信任逻辑
            int record = (data[idx] << 16) | (data[idx+1] << 8) | data[idx+2];

            while ((record & EndMask) != EndMask)
            {
                int bit = (suffix >> 15) & 1;
                
                // offset = IndexStartIndex + record * 6 + bit * 3
                int offset = IndexStartIndex + record * 6 + bit * 3;
                
                record = (data[offset] << 16) | (data[offset+1] << 8) | data[offset+2];
                suffix <<= 1;
            }

            int index = record & ComplMask;
            
            // 安全的数组访问
            var arr = _geoispArr;
            if ((uint)index < (uint)arr.Length)
            {
                return arr[index];
            }
            
            return "";
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static bool FastParseIp(string ip, out uint result)
        {
            result = 0;
            uint val = 0;
            int shift = 24;
            // 缓存Length属性
            int len = ip.Length;

            for (int i = 0; i < len; i++)
            {
                char c = ip[i];
                if (c >= '0' && c <= '9')
                {
                    val = val * 10 + (uint)(c - '0');
                }
                else if (c == '.')
                {
                    if (val > 255) return false;
                    result |= val << shift;
                    val = 0;
                    shift -= 8;
                }
                else
                {
                    return false;
                }
            }

            if (val > 255 || shift != 0) return false;
            result |= val;
            return true;
        }

        private static string FindDbPath()
        {
            string[] attempts = new[]
            {
                Path.Combine(AppDomain.CurrentDomain.BaseDirectory, DbFileName),
                Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "../data", DbFileName), 
                Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "../../../data", DbFileName), 
                "../" + DbFileName
            };

            foreach (var path in attempts)
            {
                if (File.Exists(path)) return path;
            }
            return "";
        }
    }
}
